Lib/test/test_capi/test_misc.py

```diff
@@ -401,6 +401,43 @@ def check_negative_refcount(self, code):
                          br'_Py_NegativeRefcount: Assertion failed: '
                          br'object has negative ref count')
 
+    def run_iter_api_test(self, next_func):
+        dataset = (
+            (), (1,2,3),
+            [], [1,2,3],
+        )
+
+        for input_ in dataset:
+            items = []
+            it = iter(input_)
+            while (item := next_func(it)) is not None:
+                items.append(item)
+            self.assertEqual(items, list(input_))
+
+        class Broken:
+            def __init__(self):
+                self.count = 0
+
+            def __next__(self):
+                if self.count < 3:
+                    self.count += 1
+                    return self.count
+                else:
+                    raise TypeError('bad type')
+
+        it = Broken()
+        self.assertEqual(next_func(it), 1)
+        self.assertEqual(next_func(it), 2)
+        self.assertEqual(next_func(it), 3)
+        with self.assertRaisesRegex(TypeError, 'bad type'):
+            next_func(it)
+
+    def test_iter_next(self):
+        self.run_iter_api_test(_testcapi.PyIter_Next)
```

Comment:
You can use subtests here and parameterize it

Reply:
Yeah, I thought about that, but I kept Irit's original patch for some reason.

Reply:
I'll reconsider.

Reply:
I decided to only apply subTest to the loop in `run_iter_api_test`. Thanks for the suggestion.